### PR TITLE
copr: migrate `rpn_fn` interface to use `JsonRef` and `BytesRef`

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi iskyzh@gmail.com

### What problem does this PR solve?

**Problem Summary**

This PR changes all vectorized functions to have signature like `do_something(a: Option<BytesRef>, b: Option<JsonRef>`.

**Overview**

In [RFC: Using chunk format in coprocessor framework](https://github.com/tikv/rfcs/pull/43/files), we proposed a `ChunkedVec` struct to support Chunk-based computing. We plan to:

1. Migrate current co-processor framework to use interface compatible with `ChunkedVec`. 

This is done by adding new functionalities on current `Vec<Option<T>>`. The detailed plan can be found at [Full Chunk-based Computing PR Roadmap](https://github.com/skyzh/tikv/issues/2).

This PR follows #7979. This PR, combined with #7979, defines new interface of `rpn_fn`. We'll start refactoring copr internal structures after this PR is merged.

2. Introduce `ChunkedVec`.

After migrating co-processor to have interface compatible with proposed `ChunkedVec`, we can introduce `ChunkedVec` into TiKV. Then, we may further optimize performance by merging null parameters, etc.

### What is changed and how it works?

Proposal: [RFC: Using chunk format in coprocessor framework](https://github.com/tikv/rfcs/pull/43/files)

What's Changed:

**tidb_query_codegen**

* Modify `RpnFnRefEvaluableType` an enum that accepts two kinds of types: "Ref" `&Int` and "Type" `JsonRef`
* Store `RpnFnRefEvaluableType` instead of `TypePath` in `RpnFnSignatureParam` and `VargsRpnFnSignatureParam`
* Add some workarounds which enable developers to use `JsonRef` and `BytesRef` in vectorized functions. They will be removed after we fully migrate copr to the new Chunk-based interface. This includes
    * A trait `IntoEvaluableRef` with a function `into_evaluable_ref()`. `&Json`, `&Bytes` implements this trait to return `JsonRef` and `BytesRef`. Other primitives types reference are implemented to return themselves.
    * `is_ref_type` checks if a TypePath is `JsonRef` or `BytesRef`. If yes, the generated `rpn_fn` implementation will call `into_evaluable_ref()` on arguments from copr framework, so as to transform `&Json` into `JsonRef`.
    * `is_bytes`, `is_json`, `get_vectoried_type`.
* Add `get_vargs_buf`. This function gets corresponding VARGS buffer.
* Update documentation and compile error message of `rpn_fn` macro.
* FIXME: users may use generic to bypass the limitation and use `&Json` in vectorized functions, and this can't be reported in `rpn_fn` macro expansion. This will be fixed later in two steps. After we migrated `ScalarValueRef` to new interface, we can make Rust compiler to raise compile error (copr internally won't have &Json after further migration). Then we will separate `Json` and `Bytes` logic from these generic function (ensure that generic function only accepts ref of primitive types).

**tidb_query_datatype**

* `ConvertTo<f64>`, `ConvertTo<Decimal>` and `ConvertTo<f64>` are implemented on `JsonRef`. Original `Json` implements simply calls the implementations on their reference.
* Add `IntoEvaluableRef` trait (as described above)
* Add `BytesRef` and `JsonRef`.

**tidb_query_vec_expr**

* Migrate all functions to have signature like `func(data: Option<BytesRef>)`
* In `impl_cast`, helpers like `test_as_string_helper` now accepts `T` instead of `&T`. In this way, `helper_get_cs_ref` was added to transform cs `Vec<(U, V, W)>` in original tests to be `Vec<(&U, V, W)>`, hence fitting into new generic function.
* In `impl_compare`, we removed `StringComparer` and use a single function `compare_bytes` to do string comparison.
* In `function.rs`, we added `VARG_PARAM_BUF_BYTES_REF` and `VARG_PARAM_BUF_JSON_REF` to store `BytesRef` and `JsonRef` respectively when processing vargs rpn_fn.
* FIXME: add tests for JSON compare. Meanwhile,  json compare is an example of how vectorized function bypasses the limitation that copr framework should only supply `JsonRef` to developers.

### Check List 

Tests 

- Unit test

Side effects

- Possible performance regression.
- Json comparison requires further tests. Will be fixed in next PR.

### CODE REVIEW

**Comment by skyzh** [comment]
@breeswish PTAL~

**Comment by skyzh** [comment]
Seems that copr integration test failed. Still investigating.

File: components/tidb_query_codegen/src/rpn_function.rs [hunk 1]
```diff
@@ -372,26 +372,70 @@ impl parse::Parse for RpnFnEvaluableType {
...
+
+    /// Transform new `JsonRef`-like style type to old `&Json` type.
+    ///
+    /// Note: this is a workaround for current copr framework.
+    /// After full migration, this function should be deprecated.
+    fn get_type_path(&self) -> TypePath {
+        match self {
+            Self::Type(x) => match x.path.get_ident() {
+                Some(id) => {
+                    if *id == "JsonRef" {
```
Comment (TennyZhuang) [inline components/tidb_query_codegen/src/rpn_function.rs:405]: `match *id`

File: components/tidb_query_codegen/src/rpn_function.rs [hunk 2]
```diff
@@ -372,26 +372,70 @@ impl parse::Parse for RpnFnEvaluableType {
...
+                        return parse_quote! { Json };
+                    }
+                    if *id == "BytesRef" {
+                        return parse_quote! { Bytes };
+                    }
+                    x.clone()
```
Comment (TennyZhuang) [inline components/tidb_query_codegen/src/rpn_function.rs:411]: unreachable

File: components/tidb_query_codegen/src/rpn_function.rs [hunk 3]
```diff
@@ -372,26 +372,70 @@ impl parse::Parse for RpnFnEvaluableType {
...
+                }
+                None => x.clone(),
```
Comment (TennyZhuang) [inline components/tidb_query_codegen/src/rpn_function.rs:413]: unreachable!

**Pull request merged**